libsql-sqlite3/src/vectordiskann.c

```diff
@@ -0,0 +1,345 @@
+/*
+** 2024-03-23
+**
+** Copyright 2024 the libSQL authors
+**
+** Permission is hereby granted, free of charge, to any person obtaining a copy of
+** this software and associated documentation files (the "Software"), to deal in
+** the Software without restriction, including without limitation the rights to
+** use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+** the Software, and to permit persons to whom the Software is furnished to do so,
+** subject to the following conditions:
+**
+** The above copyright notice and this permission notice shall be included in all
+** copies or substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+** FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+** COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+** IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+** CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+**
+******************************************************************************
+**
+** DiskANN for SQLite/libSQL.
+**
+** The algorithm is described in the following publications:
+**
+**   Suhas Jayaram Subramanya et al (2019). DiskANN: Fast Accurate Billion-point
+**   Nearest Neighbor Search on a Single Node. In NeurIPS 2019.
+**
+**   Aditi Singh et al (2021). FreshDiskANN: A Fast and Accurate Graph-Based ANN
+**   Index for Streaming Similarity Search. ArXiv.
+**
+**   Yu Pan et al (2023). LM-DiskANN: Low Memory Footprint in Disk-Native
+**   Dynamic Graph-Based ANN Indexing. In IEEE BIGDATA 2023.
+**
+** Here is the (internal, non-API) interface between this module and the
+** rest of the SQLite system:
+**
+**    diskAnnCreateIndex()     Create new index
+**    diskAnnDeleteIndex()     Delete existing index
+**    diskAnnOpenIndex()       Open index for operations (allocate all necessary internal structures)
+**    diskAnnCloseIndex()      Close index and free associated resources
+**    diskAnnSearch()          
+**    diskAnnInsert()          
```

Comment:
It feels like the last two functions are missing description. Was that intentional?

Reply:
I will fill them later in branch where all these functions will be introduced. Sorry for this leftover from follow up branch

Reply:
No worries. Later is good too.